### DISCUSSION TITLE
Copy & run-before

### DISCUSSION THREAD

**Discussion by ryszard-suchocki** [creativeprojects/resticprofile#125] [Q&A]
Hello,
I want to copy my repo to SMB/CIFS share. This requires mount / unmount network shared (I prefer not leaving mounted network share for everyone who uses pc/server). Is it possible or will it be possible in the future to use run-before, run-after actions in the "copy" command? I don't want to invoke mount/unmount pre/post actions with every command invoked (snapshots, stats) ("global" run-before). Most of my profiles invoke copy as backup's "run-after" action (run-after: {{configdir}}/resticprofile -c myprofile -n myprofile copy), I can then mount before and unmount after. But this way has some downsides such as an overall backup failure when the mount failed (normally only copy should fail) and I receive a "copy" notification before the "backup" notification (when everything is all right). Also, I don't want to inherit the profile as it doubles backup invoke (when scheduled - derivedprofile:backup[inherited schedule], derivedprofile:copy[new schedule] ). 

My imagination is:

copy:
  repo: myrepo
  password: myrepokey
  run-before: mount -t cifs -o "options for mount" /mount/point
  run-after: umount /mount/point

Best regards!

**Comment by creativeprojects** [comment]
I don't see any issue adding `run-before` and `run-after` on the `copy` target. It should be a fairly simple change

**Reply by jkellerer** [reply]
Was thinking that sections that can be scheduled should probably all have `run-*` targets as there may be cases to setup something special for them. (Does not mean it should be tightly bound to `schedule` but could be implemented as an action interface that the profile for globals and some sections implement)

**Reply by creativeprojects** [reply]
I agree: "keep it simple"

**Reply by ryszard-suchocki** [reply]
I will drift some from the main subject, but is it possible to disable the scheduler in derived profiles? One of the workarounds is the creation of a derived profile and adding to it a "copy" section to schedule it separately (this is the case when I don't want to copy the repo each time backup invokes). Unfortunately, the parent profile has been scheduled for the "backup" section, and I don't want to double the backups. I can set schedule: "", source: "/" ("empty" source is not doable). This causes an issue when "schedule --all" is invoked as the timer registration procedure fails.

The bigger picture - I perform VM backup (qemu/kvm) which consists of "Incr", "diff", and "full" types of backup (external snapshots - if someone is interested). I define main profile (incr, own schedule, invoked quite often), then diff derives from main (incr) which overrides "schedule" and sets env variable, full also derives from main (incr, same mechanism as in diff). I cannot simply add a schedule in the main profile for "copy" as each of diff and full will inherit the schedule. The whole resticprofile configuration that regards one specific VM is set in one .yaml file, so I can schedule all profiles at one time.

**Reply by jkellerer** [reply]
The "schedule" property is a list type (since it can have multiple schedules) and schedules are ignored when the property is an empty list. Setting it to an empty string sets it to a list with one empty entry but not an empty list. Unfortunately empty schedule strings are not ignored but cause a failure instead.

As we further have a list merge issue in V1 configs, you cannot simply set it to "`[]`" (empty list), not sure if setting it to `false` would work since the interpretation happens inside the deserialization lib.

If `false` does not work, I'd suggest to report this as bug then I can address it and filter out empty schedule definitions.

**Comment by jkellerer** [comment]
As a workaround for your current profiles (until dedicated `run-` targets are implemented), how about using:
```
run-after: {{.ConfigDir}}/resticprofile -c myprofile -n myprofile copy || echo "copy myprofile failed"
```

This will not fail the backup but write "copy myprofile failed" to the output
